### ISSUE TITLE
A concern about VAT and dropout

### ISSUE THREAD

**Issue by mingxuan-yi** [takerum/vat_tf#2]
Hi, I come back again. A question of virtual adversarial training arose to me. When using this method, we first feed the unlabeled data into the forward net to get their predictions as labels, and in the meantime, feed the noised data to get logits, then regard the KL divergence as the loss to get r_vadv. But in your codes, you applied dropout to the forward net. As far as I know, tensorflow only reuses variables and no operation (e.g. dropout) in the graph, thus the logits and the labels we get from the above procedure use different settings of dropout, i.e., the KL divergence computed may include the noises from dropout. How do you consider about it? Would it make the r_vadv not accurate enough?

**Comment by takerum** [comment]
Hi, 

For each forward pass, I set the same random seed on dropout noise. 
```h = tf.nn.dropout(h, keep_prob=FLAGS.keep_prob_hidden, seed=rng.randint(123456)) if stochastic else h```
Hence, the dropout masks in each forward  pass are the same as each other. 
You can simulate it by running the following code : 
``` 
import tensorflow as tf
sess = tf.InteractiveSession()
a = tf.random_normal([2, 2], seed=1234) # Set same random seed 
b = tf.random_normal([2, 2], seed=1234) # Set same random seed 
c = tf.equal(a,b) 
# Randomly sample Gaussian noise a, b and they are same at each run
print(sess.run([c, a, b]))
print(sess.run([c, a, b]))
```

**Comment by mingxuan-yi** [comment]
Oops, I didn't see the seed while I read that part. My apologies for my inattention. Using seeds is a really good idea. Indeed, I build another model of neural networks and applied VAT to it, it seems that via combining AT and VAT techniques, the performance increased quite a lot. (many times around 4% higher than a pure network, even I used the wrong dropout settings). In brief, thanks again, the ideas from your papers and codes are really much help.

**Comment by takerum** [comment]
No problem, and I am glad to be of help!
